src/audio_generic.cpp

```diff
@@ -62,31 +67,49 @@ void GenericAudio::BGM_Play(const std::string& file, int volume, int pitch, int
 			PlayOnChannel(BGM_Channel, file, volume, pitch, fadein);
 		}
 	}
+	UnlockMidiOutMutex();
 }
 
 void GenericAudio::BGM_Pause() {
+	LockMidiOutMutex();
 	for (auto& BGM_Channel : BGM_Channels) {
-		if (BGM_Channel.decoder) {
+		if (BGM_Channel.decoder || BGM_Channel.midiout) {
+			if (BGM_Channel.midiout) {
+				BGM_Channel.midiout->Pause();
+			}
 			BGM_Channel.paused = true;
 		}
 	}
+	UnlockMidiOutMutex();
 }
 
 void GenericAudio::BGM_Resume() {
+	LockMidiOutMutex();
 	for (auto& BGM_Channel : BGM_Channels) {
-		if (BGM_Channel.decoder) {
+		if (BGM_Channel.decoder || BGM_Channel.midiout) {
+			if (BGM_Channel.midiout) {
+				BGM_Channel.midiout->Resume();
+			}
 			BGM_Channel.paused = false;
 		}
 	}
+	UnlockMidiOutMutex();
 }
 
 void GenericAudio::BGM_Stop() {
+	LockMidiOutMutex();
 	for (auto& BGM_Channel : BGM_Channels) {
 		BGM_Channel.stopped = true; //Stop all running background music
-		LockMutex();
-		BGM_Channel.decoder.reset();
-		UnlockMutex();
+		if (BGM_Channel.midiout) {
+			BGM_Channel.midiout->Reset();
+			BGM_Channel.midiout.reset();
+		} else if (BGM_Channel.decoder) {
+			LockMutex();
```

Comment:
In this function you lock the midi mutex, and then you lock the audio mutex. Just below here,  in `BGM_PlayedOnce`, you first lock the audio mutex and then the midi mutex.

Different lock orders of multiple mutex can result in a deadlock. 

Now given that only player main thread will ever call these 2 functions, it's safe by accident.

This is another reason in favor of `std::mutex` as the standard library has facilities to lock mutiple mutex in guaranteed order.

https://en.cppreference.com/w/cpp/thread/lock